File: leetcode/move_zeroes.py
#!/usr/bin/env python3

import sys
import unittest

# number: 283
# title: Move Zeroes
# url: https://leetcode.com/problems/move-zeroes/
# section: meta
# difficulty: easy
# tags: array, two pointers, meta

# constraints
# 1 <= nums.length <= 10^4
# -2^31 <= nums[i] <= 2^31 - 1
# in-place

# solution: sliding window
# complexity
# run-time: O(n)
# space: O(1)
def move_zeroes(nums):
    left = right = 0

    for right in range(len(nums)):
        #print(f"left:{left},right:{right}")

        # find next non-zero on right
        if nums[right] == 0:
            continue

        # keep swapping if 0 on left
        if nums[left] == 0:
            nums[left],nums[right] = nums[right],nums[left]

        left += 1

class TestMoveZeroes(unittest.TestCase):
    def test_empty(self):
        nums = []
        expected = []
        move_zeroes(nums)
        self.assertEqual(nums, expected)

    def test1(self):
        nums = [0,1,0,3,12]
        expected = [1,3,12,0,0]
        move_zeroes(nums)
        self.assertEqual(nums, expected)

    def test2(self):
        nums = [0]
        expected = [0]
        move_zeroes(nums)
        self.assertEqual(nums, expected)

    def test3(self):
        nums = [1,0,1]
        expected = [1,1,0]
        move_zeroes(nums)
        self.assertEqual(nums, expected)

if __name__ == '__main__':
    sys.exit(unittest.main())

File: algomonster/move_zeroes.cpp
#include <algorithm>
#include <vector>

// tags: two pointers, sliding window

// solution: two pointers, fast & slow
// complexity:
// run-time: O(n)
// space: O(1)
void moveZeroes(std::vector<int>& v)
{
    int slow = 0;
    int fast = 0;

    while (fast < v.size()) {
        // if fast is 0, do nothing

        // if fast is not 0, swap with slow
        if (v[fast] != 0) {
            // or swap
            std::swap(v[slow], v[fast]);
            // or overwrite: don't need v[fast]
            //v[slow] = v[fast];
            ++slow;
        }

        ++fast;
    }
}